File: android/app/src/main/java/org/bitcoinppl/cove/views/ColumnMajorGrid.kt
package org.bitcoinppl.cove.views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * A grid that displays items in column-major order (top-to-bottom, then left-to-right)
 *
 * For a list [1,2,3,4,5,6] with 3 columns, displays as:
 * ```
 * 1  3  5
 * 2  4  6
 * ```
 */
@Composable
fun <T> ColumnMajorGrid(
    items: List<T>,
    modifier: Modifier = Modifier,
    numColumns: Int = 3,
    horizontalSpacing: Dp = 12.dp,
    verticalSpacing: Dp = 18.dp,
    content: @Composable (index: Int, item: T) -> Unit,
) {
    require(items.size % numColumns == 0) {
        "Item count (${items.size}) must be divisible by $numColumns"
    }
    val itemsPerColumn = items.size / numColumns

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(horizontalSpacing),
    ) {
        repeat(numColumns) { col ->
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(verticalSpacing),
            ) {
                repeat(itemsPerColumn) { row ->
                    val index = col * itemsPerColumn + row
                    if (index < items.size) {
                        content(index, items[index])
                    }
                }
            }
        }
    }
}

File: ios/Cove/Views/ColumnMajorGrid.swift
//
//  ColumnMajorGrid.swift
//  Cove
//
//  Created by Praveen Perera on 1/16/26.
//

import SwiftUI

/// A grid that displays items in column-major order (top-to-bottom, then left-to-right)
///
/// For a list [1,2,3,4,5,6] with 2 columns, displays as:
/// ```
/// 1  4
/// 2  5
/// 3  6
/// ```
struct ColumnMajorGrid<Item, Content: View>: View {
    let items: [Item]
    let numberOfColumns: Int
    let spacing: CGFloat
    let content: (Int, Item) -> Content

    init(
        items: [Item],
        numberOfColumns: Int = 2,
        spacing: CGFloat = 12,
        @ViewBuilder content: @escaping (Int, Item) -> Content
    ) {
        self.items = items
        self.numberOfColumns = numberOfColumns
        self.spacing = spacing
        self.content = content
    }

    private var itemsPerColumn: Int {
        precondition(
            items.count % numberOfColumns == 0,
            "Item count (\(items.count)) must be divisible by \(numberOfColumns)"
        )
        return items.count / numberOfColumns
    }

    var body: some View {
        HStack(alignment: .top, spacing: spacing) {
            ForEach(0 ..< numberOfColumns, id: \.self) { col in
                VStack(spacing: spacing) {
                    ForEach(0 ..< itemsPerColumn, id: \.self) { row in
                        let index = col * itemsPerColumn + row
                        content(index, items[index])
                    }
                }
                .frame(maxWidth: .infinity)
            }
        }
    }
}
